### DISCUSSION TITLE
There is some way to detect collision between objects?

### DISCUSSION THREAD

**Discussion by F3RN4NDOJF3RN4ND3Z** [StereoKit/StereoKit#237] [Q&A] [answered]
Hi I am amazed with this framework it has been really easy to start unsing it thanks for the hard work you are doin. I was wondering if it is possible to trigger an animation in a model when another model collides with a plane on it, because the examples like the painting app shows ui reacting with hands when it touches an specific color, but I dont see examples about models reacting with other models in a scene.

Thanks

**Comment by F3RN4NDOJF3RN4ND3Z** [comment, marked as answer]
I start with Drawing some boxes to see the bounds of the models and try intersecting the ray from the poses I wanted to intersect. I dont know what I am doing wrong but the box is not placed at the center of the model

this are my Poses
` Pose syrengePose = new Pose(0f, -0.25f, -0.4f, Quat.FromAngles(0,90,0));`
`Pose vialPose = new Pose(-0.2f, -0.25f, -0.4f,Quat.Identity);`

And this is how I am Drawing the cubes arround the models
`
      Mesh.Cube.Draw(Material.UIBox, Matrix.TRS(vialPose.position, vialPose.orientation,vial.Bounds.dimensions));
      Mesh.Cube.Draw(Material.UIBox, Matrix.TRS(syrengePose.position, syrengePose.orientation, syrenge.Bounds.dimensions));
`

here is the ray intersect test
`bool inVial=syrengePose.Ray.Intersect(vial.Bounds, out intersectionPoint)`

but this is not hiting iun any direction

here is an image of what is going on
![image](https://user-images.githubusercontent.com/12785550/143284529-0c1b4a72-ac82-416f-b57c-ca71b0c32a1e.png)

**Reply by maluoi** [reply]
It's probably important to mention here that Ray intersection math needs to be done in the coordinate space relative to your collision object. You are on the right path here though, right now Ray intersections are probably the best method of doing collision testing. Here's some modifications to your sample code that may hopefully be illuminating?

```csharp
Pose syringePose = new Pose(0f, -0.25f, -0.4f, Quat.FromAngles(0,90,0));
Matrix vialTransform = Matrix.T(-0.2f, -0.25f, -0.4f);

// Make the Ray relative to the vial
Matrix vialInvTransform = vialTransform.Inverse;
Ray syringeRay = vialInvTransform.Transform(syringePose.Ray);

// Your original Draw code needed to also account for Bounds.center
Mesh.Cube.Draw(Material.UIBox, Matrix.TS(vial.Bounds.center, vial.Bounds.dimensions) * vialTransform); 

bool inVial = syringeRay.Intersect(vial.Bounds, out intersectionPoint)
```
Just as a note, I haven't tested this code specifically, but hopefully it conveys what you're looking for! :) You can see an additional example of this that definitely works in the docs over on [this page](https://stereokit.net/Pages/Reference/Ray/Intersect.html#ray-mesh-intersection)!

**Reply by F3RN4NDOJF3RN4ND3Z** [reply]
I finally make the bounding boxes draw in their place but I am still having issues insersecting with the ray, I also try to use contains for the bounds but again it seems to be missplaced

here is how i can draw the cubes in its place.

```C#
 Mesh.Cube.Draw(Material.UIBox, Matrix.TS(vial.Bounds.center, vial.Bounds.dimensions) * vialPose.ToMatrix());
 Mesh.Cube.Draw(Material.UIBox, Matrix.TS(syrenge.Bounds.center, syrenge.Bounds.dimensions) * syrengePose.ToMatrix());
```
then I draw a line  and the axis to the pose between the two models to test if they are relative to each other it is supose to appear at the center of each model as in the example of the cubes
![20211125_074347_MixedReality](https://user-images.githubusercontent.com/12785550/143444165-7489ef1c-39c2-48d9-8bf4-2ceeb0664480.jpg)

I start to think that something is wrong with the models that I imported

**Reply by F3RN4NDOJF3RN4ND3Z** [reply]
I got it I have to work with this matrixes instead of the Pose directly
```C#
      Matrix vialPoseTransform = Matrix.TS(vial.Bounds.center, vial.Bounds.dimensions) * vialPose.ToMatrix();
      Matrix syrengePoseTransform = Matrix.TS(syrenge.Bounds.center, syrenge.Bounds.dimensions) * syrengePose.ToMatrix();
```

that make the trick 
how can I draw the ray? it is possible to see where the ray is comming from?

**Reply by maluoi** [reply]
A `Ray` can be drawn with `Lines.Add`, there's an overload that takes a Ray as an argument already :) You'll want to draw the `Ray` in world space, and not the `Ray` that's been transformed to be relative.

One thing to watch out with for `Model`s, is that it's kinda easy to get the meshes off-center in the original modeling tool! If your `Model`s are really off center, and you rely on that a lot, you may find it worth your time to correct the original model asset. It can certainly simplify some of the code! But if that's not an option, the `Model.Bounds.center` does indicate the offset itself, and you can transform that offset into something usable like this:

```csharp
Vec3 vialCenterPos = vialPose.ToMatrix().Transform(vial.Bounds.center);
```
